tests/rados/test_osd_full.py

```diff
@@ -51,272 +57,640 @@ def run(ceph_cluster, **kw):
+                if "stray daemon" in health_detail:
+                    cephadm.shell(args=["ceph orch restart mgr"])
+                for osd_id in a_set:
+                    assert f"osd.{osd_id} {osd_warn}" in health_detail
+                assert f"pool '{pool_name}' {pool_warn}" in health_detail
+                break
+            except AssertionError:
+                time.sleep(10)
+                if datetime.datetime.now() >= timeout_time:
+                    raise
+
+    if config.get("pg_autoscaling"):
+        doc = (
+            "\n#CEPH-83571715"
+            "\n Verify cluster behaviour when OSDs are nearfull, backfill-full, and completely full."
+            "\n Autoscale PGs, ensure proper backfilling and rebalancing"
+            "\n\t 1. Create replicated pool with single PG and autoscaling disabled"
+            "\n\t 2. Retrieve the acting set for the pool"
+            "\n\t 3. Calculate the amount of data to be written to pools in"
+            " order to trigger near-full, backfill-full, and completely full OSD health warnings"
+            "\n\t 4. Perform IOPS using rados bench to fill up the pool up till nearfull,"
+            " backfillfull and completely full capacity"
+            "\n\t 5. Verify the number of objects written to the pool and the cluster"
+            " health warning after each IOPS iteration"
+            "\n\t 6. Perform IOPS again and ensure IOPS was unsuccessful as OSDs are full"
+            "\n\t 7. Increase the set-full-ratio to a higher value and set-backfillfull-ratio"
+            " to a value higher than the previous value of set-full-ratio"
+            "\n\t 8. Ensure that completely full and backfillfull warning disappear and "
+            "nearfull warning remains"
+            "\n\t 9. Enable autoscaling and observe the acting set OSD's 'used capacity' decreasing"
+            "\n\t 10. Eventually nearfull warning should disappear, wait for PGs to achieve active+clean state"
+        )
+
+        log.info(doc)
+        config = config["pg_autoscaling"]
+        log.info("Running pg rebalance with full OSDs test case")
+        pool_target_configs = config["pool_config"]
+
+        for entry in pool_target_configs.values():
+            try:
+                rados_obj.configure_pg_autoscaler(**{"default": "off"})
+                pool_name = entry["pool_name"]
+                assert create_pool_per_config()
+                # retrieving the size of each osd part of acting set for the pool
+                acting_set = rados_obj.get_pg_acting_set(pool_name=pool_name)
+                for osd_id in acting_set:
+                    osd_df_stats = rados_obj.get_osd_df_stats(
+                        tree=False, filter_by="name", filter=f"osd.{osd_id}"
+                    )
+                    osd_sizes[osd_id] = osd_df_stats["nodes"][0]["kb"]
+
+                primary_osd_size = osd_sizes[acting_set[0]]
+                for osd_id in acting_set:
```

Comment:
OSDs can be part of multiple PGs, the total capacity of the OSD will not change due to this. This check is to ensure that the OSDs are of same size so when Primary OSD is filled till a certain %, other OSDs part of the acting set are also close if not more.
The test has a prerequisite that OSDs used for writing data have minimal/negligible data already existing in them as the calculation has been done as per this assumption.

Reply:
Also this part of the code has not been added/modified with this PR, it was just moved from another test module to this module. If changes are necessary, it can be taken up as a separate PR

Reply:
Oh, ok. Thank you for clarification